File: 022-括号生成/022.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> vec;
        gen("",vec,n,n);
        return vec;
    }
    void gen(string sublist, vector<string> &vec,int left,int right){
        if (left == 0 && right == 0){
            vec.push_back(sublist);
            return;
        }
        if (left > 0) {
            gen(sublist+ "(", vec, left-1, right);
        }
        if (right > left) {
            gen(sublist+ ")", vec, left, right-1);
        }
    }
};

File: 022-括号生成/022.java
class Solution {
  private List<String> res;
  public List<String> generateParenthesis(int n) {
      res = new ArrayList<String>();
      generate(0, 0, n, "");
      return res;
  }
  public void generate(int left, int right, int n, String s){
      // terminator
      if (left == n && right == n){
          res.add(s);
          return;
      }
      
      // process left right
      
      // drill down
      if(left < n){
          generate(left + 1, right,n, s + "(");
      }
      if(left > right){
          generate(left, right + 1, n, s + ")");
      }
      
      // reverse states
  }
}